### PR TITLE
8290417: CDS cannot archive lamda proxy with useImplMethodHandle

### PR DESCRIPTION
**Pull request by iklam**
Since the impact of this bug can be big, and the fix is simple and low risk (just avoids doing optimizations), I'd like to put this into JDK 19 before RDP2 (Jul 21).

CDS cannot handle Lambda proxy classes that are generated in the useImplMethodHandle mode. This could happen with classfiles generated by JDK 8 or JDK 11 that access protected methods in a base class from a different package. E.g.,

```
class pkg1.Base {
    protected void doit(String s) {
        System.out.println(s);
    }
}
class pkg2.Child  extends pkg1.Base {
    void test() {
        Optional<String> opt = Optional.of("foo");
        opt.ifPresent(this::doit);
    }
}
```

More details of useImplMethodHandle can be found here: https://github.com/openjdk/jdk/blob/522b65743ca10fcba0a27d25b8fa11319999e228/src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java#L183-L191

More details about the condition that triggers the error can be found in the test file Child.jcod.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290417](https://bugs.openjdk.org/browse/JDK-8290417): CDS cannot archive lamda proxy with useImplMethodHandle

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5ce2cbd9](https://git.openjdk.org/jdk19/pull/146/files/5ce2cbd98f55fa47941e7e04dab78ab0764fff28)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [7122e367](https://git.openjdk.org/jdk19/pull/146/files/7122e3672cd8cab780b0e0f9343ff43bca7c540a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk19 pull/146/head:pull/146` \
`$ git checkout pull/146`

Update a local copy of the PR: \
`$ git checkout pull/146` \
`$ git pull https://git.openjdk.org/jdk19 pull/146/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 146`

View PR using the GUI difftool: \
`$ git pr show -t 146`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk19/pull/146.diff">https://git.openjdk.org/jdk19/pull/146.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/cds/appcds/LambdaWithUseImplMethodHandle.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8290417
+ * @summary CDS cannot archive lamda proxy with useImplMethodHandle
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/LambdaWithUseImplMethodHandle.java:28]: typo: lamda -> lambda

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8290417
+ * @summary CDS cannot archive lamda proxy with useImplMethodHandle
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java:28]: typo: lamda -> lambda

**Review by dholmes-ora — APPROVED** [review]
Looks good!

Thanks.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * @build pkg2.Child
+ * @build LambdaWithUseImplMethodHandleApp
+ * @build sun.hotspot.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller -jar test.jar pkg1.BaseWithProtectedMethod pkg2.Child LambdaWithUseImplMethodHandleApp
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xbootclasspath/a:. DynamicLambdaWithUseImplMethodHandle
+ */
+
+import jdk.test.lib.cds.CDSOptions;
+import jdk.test.lib.cds.CDSTestUtils;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java:41]: Extra import statements?

**Review by calvinccheung — APPROVED** [review]
Looks good. Just couple of nits.

**Review by mlchung — APPROVED** [review]
Looks fine.    I agree that this fix is low risk to disable archiving the lambda proxy in the case when passing a live method handle for invocation.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java (hunk 3 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/DynamicLambdaWithUseImplMethodHandle.java]: Fixed.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @dholmes-ora, @calvinccheung, @mlchung for the review.
Passed tiers 1 - 5.
/integrate

**Pull request closed without merging**